### PR TITLE
feat(web api): support request.clone()

### PR DESCRIPTION
**Pull request by zhiyuang**
Support request.clone

fix https://github.com/oven-sh/bun/issues/1381

### CODE REVIEW

File: src/bun.js/webcore/response.zig [hunk 1]
```diff
@@ -4200,7 +4200,7 @@ pub const InternalBlob = struct {
     pub fn toJSON(this: *@This(), globalThis: *JSC.JSGlobalObject) JSValue {
         const str_bytes = ZigString.init(this.bytes.items).withEncoding();
         const json = str_bytes.toJSONObject(globalThis);
-        this.deinit();
+        // this.deinit();
```
Comment (Jarred-Sumner) [inline src/bun.js/webcore/response.zig]: Why is this commented out? Wouldn't it cause the memory to never be freed?

File: src/bun.js/api/server.zig [hunk 2]
```diff
@@ -1035,16 +1037,16 @@ fn NewRequestContext(comptime ssl_enabled: bool, comptime debug_mode: bool, comp
...
 
-                this.request_js_object = null;
+                // this.request_js_object = null;
                 defer request_js.ensureStillAlive();
                 defer JSC.C.JSValueUnprotect(this.server.globalThis, request_js.asObjectRef());
                 // User called .blob(), .json(), text(), or .arrayBuffer() on the Request object
                 // but we received nothing or the connection was aborted
                 if (request_js.as(Request)) |req| {
                     // the promise is pending
-                    if (req.body == .Locked and req.body.Locked.action != .none and req.body.Locked.promise != null) {
```
Comment (Jarred-Sumner) [inline src/bun.js/api/server.zig:1045]: Was this a bug? This feels unrelated

File: src/bun.js/api/server.zig [hunk 3]
```diff
@@ -658,6 +658,8 @@ fn NewRequestContext(comptime ssl_enabled: bool, comptime debug_mode: bool, comp
         request_body_content_len: usize = 0,
         sink: ?*ResponseStream.JSSink = null,
         byte_stream: ?*JSC.WebCore.ByteStream = null,
+        request_body_waiting_for_data: ?*JSC.WebCore.Body.Value = null,
```
Comment (Jarred-Sumner) [inline src/bun.js/api/server.zig:661]: What ensures this `Body.Value` pointer is alive for the duration of the request?

File: src/bun.js/webcore/streams.zig [hunk 4]
```diff
@@ -118,6 +118,33 @@ pub const ReadableStream = struct {
         ReadableStream__cancel(this.value, globalThis);
     }
 
+    pub fn tee(this: *const ReadableStream, globalThis: *JSGlobalObject) [2]ReadableStream {
+        JSC.markBinding(@src());
+        this.value.unprotect();
+        var branches = ReadableStream__tee(this.value, globalThis);
```
Comment (Jarred-Sumner) [inline src/bun.js/webcore/streams.zig:124]: Please move this logic to `ReadableStream__tee ` and have `ReadableStream__tee` return the two ReadableStream `JSValue` directly

That will let us skip creating these empty ReadableStream objects (which are expensive to create) and skip the array iterator stuff. You can use `getDirectIndex` on the `JSArray` object

**Review by Jarred-Sumner — CHANGES_REQUESTED** [review]
Thank you for this and for writing tests!

A few comments

File: src/bun.js/webcore/response.zig (hunk 1 above)
Comment (zhiyuang) [reply src/bun.js/webcore/response.zig]: oh, this should be added back

File: src/bun.js/api/server.zig (hunk 2 above)
Comment (zhiyuang) [reply src/bun.js/api/server.zig:1045]:
``` javascript
export default {
  async fetch(request) {
    console.log(".json()", request.json());
    return new Response("Ok");
  }
}
```

In this case, it would output TypeError: Request aborted, if not commenting it out.

File: src/bun.js/webcore/streams.zig (hunk 4 above)
Comment (zhiyuang) [reply src/bun.js/webcore/streams.zig:124]: OK

File: src/bun.js/api/server.zig (hunk 3 above)
Comment (zhiyuang) [reply src/bun.js/api/server.zig:661]: it is alive when .json() is triggered, and waiting for request body data. The purpose of this is to avoid using `request_js_object`'s body value dirrectly, which would make its internal status not right sometimes.

File: src/bun.js/bindings/webcore/ReadableStream.cpp [hunk 5]
```diff
@@ -136,8 +136,16 @@ JSC::JSValue ReadableStream::tee()
     arguments.append(JSC::jsBoolean(true));
     ASSERT(!arguments.hasOverflowed());
     auto returnedValue = invokeReadableStreamFunction(lexicalGlobalObject, privateName, JSC::jsUndefined(), arguments);
+
+    JSArray* array = JSC::jsDynamicCast<JSArray*>(returnedValue.value_or(JSC::jsUndefined()));
+
+    VM& vm = lexicalGlobalObject.vm();
+    JSC::JSObject* object = JSC::constructEmptyObject(&lexicalGlobalObject, lexicalGlobalObject.objectPrototype(), 2);
+
+    object->putDirect(vm, JSC::Identifier::fromString(vm, "branch1"_s), array->getDirectIndex(&lexicalGlobalObject, 0), 0);
```
Comment (Jarred-Sumner) [inline src/bun.js/bindings/webcore/ReadableStream.cpp:145]: We can skip creating this object

Instead, either return 
```c++
EncodedJSValue[2]{  JSValue::encode(array->getDirectIndex(&lexicalGlobalObject, 0)), JSValue::encode(array->getDirectIndex(&lexicalGlobalObject, 0)) };
```

or make it a pointer we pass to this function

File: test/bun.js/serve.test.ts [hunk 6]
```diff
@@ -139,6 +139,8 @@ describe("web api - request", () => {
         async fetch(req) {
           const body = await req.clone().json();
           expect(JSON.stringify(body)).toBe(JSON.stringify(requestBody));
+          const body2 = await req.json();
+          expect(JSON.stringify(body2)).toBe(JSON.stringify(requestBody));
```
Comment (Jarred-Sumner) [inline test/bun.js/serve.test.ts]: now that `.toEqual` is implemented, we could use that instead

File: test/bun.js/serve.test.ts [hunk 7]
```diff
@@ -214,6 +218,37 @@ describe("web api - request", () => {
     );
   });
 
+  it("request.clone not allowed when body is used", async () => {
```
Comment (Jarred-Sumner) [inline test/bun.js/serve.test.ts:221]: nice

File: src/bun.js/bindings/webcore/ReadableStream.cpp (hunk 5 above)
Comment (zhiyuang) [reply src/bun.js/bindings/webcore/ReadableStream.cpp:145]: then how to get value in Zig part, still using `arrayIterator`?

**Comment by Jarred-Sumner** [comment]
will merge after i manually test

**Comment by JAD3N** [comment]
This problem seems to be the only thing holding me back from using Bun. Is there anything I can do to help move this pull request along?

**Comment by danthegoodman1** [comment]
This has caused my to remove bun for tsx in 3 projects already, is there a projection on when this will be ready?

**Comment by danthegoodman1** [comment]
Why was this closed?

**Comment by Jarred-Sumner** [comment]
@danthegoodman1 @billywhizz will have a better implementation today
